lib/apiservers/engine/backends/convert/stats.go

```diff
@@ -0,0 +1,266 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/docker/docker/api/types"
+
+	"github.com/vmware/vic/lib/portlayer/metrics"
+)
+
+// ContainerStats encapsulates the conversion of VMMetrics to
+// docker specific metrics
+type ContainerStats struct {
+	config         ContainerStatsConfig
+	dockerStats    *types.StatsJSON
+	currentMetrics *metrics.VMMetrics
+	totalVCHMhz    uint64
+	dblVCHMhz      uint64
+	preTotalMhz    uint64
+
+	// reader/writer for stream
+	reader *io.PipeReader
+	writer *io.PipeWriter
+}
+
+type ContainerStatsConfig struct {
+	Ctx         context.Context
+	Cancel      context.CancelFunc
+	Out         io.Writer
+	ContainerID string
+	Memory      int64
+	Stream      bool
+	VchMhz      int64
+}
+
+type InvalidOrder struct {
+	current  time.Time
+	previous time.Time
+}
+
+func (iso InvalidOrder) Error() string {
+	return fmt.Sprintf("The current sample time (%s) is before the previous time (%s)", iso.current, iso.previous)
+}
+
+// NewContainerStats will return a new instance of ContainerStats
+func NewContainerStats(config ContainerStatsConfig) *ContainerStats {
+	return &ContainerStats{
+		config:      config,
+		dockerStats: &types.StatsJSON{},
+		totalVCHMhz: uint64(config.VchMhz),
+		dblVCHMhz:   uint64(config.VchMhz * 2),
+	}
+}
+
+// Stop will clean up remaining conversion resources
+func (cs *ContainerStats) Stop() {
+	if cs.reader != nil && cs.writer != nil {
+		cs.reader.Close()
+		cs.writer.Close()
+	}
+}
+
+// Listen will listen for new metrics from the portLayer, covert to docker format
+// and encode to the configured Writer.  The returned PipeWriter is the source of
+// the vic metrics that will be transformed to docker stats
+func (cs *ContainerStats) Listen() *io.PipeWriter {
+	// TODO: could split decode / encode into separte funcs -- would provide for easier
+	// unit testing
+
+	// we already are listening
+	if cs.reader != nil {
+		return nil
+	}
+
+	cs.reader, cs.writer = io.Pipe()
+
+	dec := json.NewDecoder(cs.reader)
+	doc := json.NewEncoder(cs.config.Out)
+
+	// channel to transfer metric from decoder to encoder
+	// closed w/in the decoder
+	metric := make(chan metrics.VMMetrics)
+
+	// signal to decoder / encoder that we are done
+	finished := make(chan struct{})
+
+	var vmm metrics.VMMetrics
+	var previousStat *types.StatsJSON
+
+	go func() {
+		for {
+			select {
+			case <-cs.config.Ctx.Done():
+				close(finished)
+			case <-finished:
+				close(metric)
+				cs.Stop()
+				return
+			default:
+				for dec.More() {
+					err := dec.Decode(&vmm)
+					if err != nil {
+						log.Errorf("container metric decoding error for container(%s): %s", cs.config.ContainerID, err)
+						cs.config.Cancel()
```

Comment:
It shouldn't matter -- it would go to the conversion stage and wouldn't be able to be converted to a docker stat and nothing would happen.  Plus the context cancel would probably catch it before it went to far.  

If you feel strongly that it should be a return after that I can do that...